### PR TITLE
fix(api): stop the DEBUG response check turning a good response into a 500

### PR DESCRIPTION
**Pull request by andrewm4894**
## Problem

An endpoint that declares a `DataclassSerializer` response returns 500 in local development, after writing its row. The caller cannot tell a successful write from a failed one, and the row is already there.

- `@validated_request` re-parses every response through its declared serializer when `DEBUG` is on.
- `serializer.is_valid()` converts only DRF's own `ValidationError` into a return value.
- A `DataclassSerializer` rebuilds its dataclass in `to_internal_value`, where read-only fields are already stripped, so a pydantic dataclass raises for a response that is perfectly valid.
- That exception escapes `is_valid()` and fails the request the check was only inspecting.

Production is unaffected: the block runs under `DEBUG` or explicit `strict_response_validation`. The test suite never sees it either, because tests run with `DEBUG=False`.

## Changes

- An endpoint whose declared serializer cannot parse its own response now returns that response unchanged, and logs a warning.
- The warning names the view, the status code and the serializer class, matching how a field mismatch is already reported.
- `strict_response_validation` still raises, so the strict callers keep the behavior they opted into.

Nothing user-visible changes in production. The fix only affects requests served with `DEBUG` on, or with strict validation set.

## How did you test this code?

`posthog/api/test/test_mixins.py` gains two cases. Neither is covered today, because every existing case uses a serializer that parses its own output.

- A response serializer that raises a non-DRF exception returns the original response.
- The same serializer under `strict_response_validation` still raises.

Not checked: no manual run against a live server in this session. The 500 was measured earlier on a stable backend against the autoresearch endpoints, same payload and endpoint either side of the fix: 500 before, 201 after.

## Automatic notifications

- [ ] Publish to changelog?

## Docs update

None.

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Claude Code (Opus 5) found this while running the autoresearch training loop locally: a sandbox agent was probing payload shapes to find one that "worked", because every write it made returned 500 and still succeeded. All six autoresearch `DataclassSerializer`s hit the same path, so it was the decorator, not the product.

The fix is cut fresh onto master from the autoresearch integration branch, ahead of the autoresearch PR stack, because it is core infrastructure that no autoresearch layer should gate.

Skills invoked: `/writing-pr-descriptions`.

### CODE REVIEW

**Review by gantoine — APPROVED** [review]
looks good but i'd want a second dev to approve since this is a "dangerous" change if its wrong

File: posthog/api/test/test_mixins.py [hunk 1]
```diff
@@ -200,6 +210,60 @@ def mock_endpoint(view_self, request):
         assert response.status_code == status.HTTP_200_OK
         assert response.data["wrong_field"] == "value"
 
+    def test_response_serializer_that_raises_while_parsing_logs_warning(self):
+        """A response serializer raising during parsing should warn, not fail the request"""
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/api/test/test_mixins.py]: Agreed, removed both docstrings in 57c843c29fc. The older tests in this file have them, but the repo rule is no docstrings in Python tests, so the new ones follow the rule rather than the file.

File: posthog/api/test/test_mixins.py [hunk 2]
```diff
@@ -30,6 +30,16 @@ class ErrorResponseSerializer(serializers.Serializer):
     detail = serializers.CharField()
 
 
+class RaisingResponseSerializer(serializers.Serializer):
+    """Stands in for a DataclassSerializer, which rebuilds its dataclass while parsing and so can
+    raise a non-DRF error for a response that is perfectly valid."""
+
+    value = serializers.CharField()
+
+    def to_internal_value(self, data):
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/api/test/test_mixins.py]: Fixed in 57c843c29fc: `def to_internal_value(self, data: object) -> NoReturn`. `NoReturn` is the accurate type since the override only ever raises.

File: posthog/api/mixins.py [hunk 3]
```diff
@@ -239,7 +239,29 @@ def wrapper(self, request: Request, *args, **kwargs) -> Response:
...
+                    if strict_response_validation:
+                        raise
+                    logger.warning(
+                        "Response serializer could not parse the response it declared for status code "
+                        f"{status_code} in the responses parameter of the @validated_request decorator. "
+                        "The response was returned unchanged; check the declared serializer.",
+                        view_func=view_func.__name__,
+                        status_code=status_code,
+                        serializer_class=serializer_class.__name__,
+                        error=str(exc),
```
[the remark this answers is not in the document]
Comment (andrewm4894) [reply posthog/api/mixins.py:256]: Keeping `error=str(exc)` here, for two reasons.

This line only runs under `DEBUG`. The enclosing block is gated on `strict_response_validation or settings.DEBUG`, and the strict branch re-raises before reaching the log, so production never executes it. Under `DEBUG`, Django already renders full tracebacks with locals for any error, so this log adds no exposure that does not already exist in that mode.

The message is also the diagnostic. For a pydantic dataclass it says which field was missing or wrong, which is exactly what a developer needs to fix the declared serializer. The exception type alone (`ValidationError`) would not tell them what to change. The sibling warning just below this one already logs `validation_errors=serialized.errors` for the same reason, so this keeps the two paths consistent.
